stream_alert/parsers.py

```diff
@@ -65,24 +65,90 @@ def __init__(self, data, schema, options):
 
     @abstractmethod
     def parse(self):
-        """Main parser method to be overridden by all Parser classes"""
+        """Main parser method to be overridden by all Parser classes
+
+        Returns:
+            A list of parsed records
+        """
         pass
 
 @parser
 class JSONParser(ParserBase):
     __parserid__ = 'json'
 
+    def __init__(self, *args):
+        super(JSONParser, self).__init__(*args)
+        self.nested = False
+        self.nested_keys = []
+
+    def _get_schema(self):
+        """Return the schema, handle nested json types"""
+        schema = self.schema
+        if self.nested:
+            for key in self.nested_keys:
+                schema = schema.get(key)
+            return schema[0]
+
+        return schema
+
+    def _key_check(self, json_records):
+        """Verify the declared schema matches the json payload
+
+        If keys do not match per the schema, records are removed from the
+        passed in json_records list
+        """
+        schema = self._get_schema()
+        schema_keys = set(schema.keys())
+
+        for json_record in json_records:
+            json_keys = set(json_record.keys())
+            if json_keys == schema_keys:
+                for key, key_type in schema.iteritems():
+                    # If the value is a map of defined key/value pairs
+                    if isinstance(key_type, dict) and key_type != {}:
+                        # subkey check
+                        if set(json_record[key].keys()) != set(schema[key].keys()):
+                            json_records.remove(json_record)
+            else:
+                logger.debug('JSON Key mismatch: %s vs. %s', json_keys, schema_keys)
+                json_records.remove(json_record)
+
+    def _parse_records(self, json_payload):
+        """Iterate over a json_payload to determine if nested records exist
+
+        Args:
+            json_payload: A dict of the parsed json data
+            schema: A dict of a log type's schema
+
+        Returns:
+            A list of dict JSON payloads
+        """
+        json_records = []
+
+        if len(json_payload) == 1:
+            for key, val in json_payload.iteritems():
+                self.nested = True
+                self.nested_keys.append(key)
+                if isinstance(val, list):
+                    for record in val:
+                        json_records.append(record)
```

Comment:
one thing i was wondering if is it would be worth merging the fields from the enveloping event with the nested events? for cloudwatch log subscriptions, the envelope fields would be `logGroup` `logStream` `owner` and `subscriptionFilters`. as is, it seems like we would lose those fields as we discard the enveloping event